Visualize body particle faces and ellipsoid particles

akohlmey opened the pull request:
**Summary**

This pull request adds the missing code to enable drawing body particle faces for body styles `rounded/polyhedron` and `rounded polygon`. The existing code is somewhat refactored to be simpler and more like that for tri/line particles. Also some minor issues from previous `dump image` related pull requests are addressed. In addition, this pull request adds support for rendering ellipsoid particles by approximating the ellipsoid as a mesh of triangles.
Furthermore, also an arrow graphics object (constructed from triangles and a cylinder) was implemented and added to fix graphics.

![image 001000](https://github.com/user-attachments/assets/a8838216-fbc5-486f-b1cb-596b141dea08)
<img width="600" height="600" alt="myimage-000030000" src="https://github.com/user-attachments/assets/8ce9d093-79dc-47fa-b967-543adfd0ae83" />

<img width="800" height="800" alt="ellipsoid-optimized" src="https://github.com/user-attachments/assets/f26e44c8-f2f0-4ffe-944c-09746ad26c68" />

<img width="800" height="800" alt="arrows-test" src="https://github.com/user-attachments/assets/b603d515-211d-4d9d-8731-b8d985d979f2" />

**Related Issue(s)**

The changes in this pull request are further extended in PR #4819 and parts of it refactored in PR #4820. Thus any changes about the details of the implementation are better done in PR #4820 and comments added there.

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Artificial Intelligence (AI) Tools Usage**

Some of the low-level graphics code was prototyped by Claude Sonnet (through Perplexity.ai) and Google Gemini. It has been significantly refactored, though.

**Backward Compatibility**

N/A

**Implementation Notes**

The body style `image()` function now increments the number of graphics elements it computed.
Some invariant calculations were moved outside the loops.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system

akohlmey committed:
2d is determined by the dimension command not boundary

akohlmey committed:
cosmetic changes

akohlmey committed:
add preliminary support for displaying body faces to dump image

akohlmey committed:
Merge branch 'develop' into visualize-body-faces

akohlmey committed:
Merge branch 'develop' into visualize-body-faces

akohlmey committed:
adjust image data allocation to accommodate all required elements

akohlmey committed:
support drawing bodies as filled objects and wireframes

akohlmey committed:
update docs/Howto for added functionality

akohlmey committed:
simplify

akohlmey committed:
export body, line, and tri flag setting in library interface

akohlmey committed:
unwrap some lines

akohlmey committed:
cosmetic

akohlmey committed:
refactor dump image support in body style rounded polygon

akohlmey committed:
update docs

akohlmey committed:
small corrections

akohlmey committed:
shift triangles when displaying both edges and faces so they are flush

akohlmey committed:
remove dead code

akohlmey committed:
silence compiler warnings

akohlmey committed:
Merge branch 'collected-small-fixes' into visualize-body-faces

akohlmey committed:
refactor ellipsoid construction for more code sharing and less redundancy

akohlmey committed:
fix off-by-one bug

akohlmey committed:
fix indexing bug

akohlmey committed:
add support for rendering ellipsoids

akohlmey committed:
add docs for ellipsoid support in dump image

akohlmey committed:
spelling

akohlmey committed:
use upper case for constants

akohlmey committed:
make docs consistent

akohlmey committed:
since we removed std::move() we can move back to consistent ordering

akohlmey committed:
small grammar fix

akohlmey committed:
make handling of color and opacity consistent and document dump_modify additions

ndtrung81 commented on src/BODY/body_rounded_polyhedron.cpp:
```diff
@@ -601,64 +593,189 @@ void BodyRoundedPolyhedron::output(int ibonus, int m, double *values)
+  int nelements = 0;
+  AtomVecBody::Bonus *const bonus = &avec->bonus[ibonus];
+  const double *const x = atom->x[bonus->ilocal];
+  const double diam = (flag1 <= 0.0) ? 2.0 * enclosing_radius(bonus) : flag1;
```
`const double diam = (flag1 <= 0.0) ? 2.0 * rounded_radius(bonus) : flag1;`

akohlmey committed:
increment numerical version for development branch so it is larger than the release number

akohlmey committed:
bug fix from Trung

akohlmey committed:
work around test failures due to incrementing numeric version for develop

akohlmey committed:
ensure the progress bar indicator not of zero length to avoid artifacts

akohlmey committed:
make documentation of color and transparency for wall fixes consistent

akohlmey committed:
fix argument parsing bug

akohlmey committed:
lower resolution for fewer rounding artifacts

akohlmey committed:
change order of triangle corners for correct orientation of the surface normal

akohlmey committed:
draw an uncapped cylinder on top of the tiled object for a smoother outside

akohlmey committed:
add a small overlap when placing triangles on cones and cylinder surface to reduce atifacts from rounding issues

akohlmey committed:
add arrow graphics object and refactor vector functions

akohlmey committed:
correct small documentation issues

akohlmey committed:
remove dead code

akohlmey committed:
improve image quality for ellipsoids and reduce artifacts when using edges and trianges together

akohlmey committed:
make formatting consistent and add some comments

akohlmey committed:
skip the complexity of creating triangle mesh for real spheres

akohlmey committed:
fix typo

sjplimp commented:
@akohlmey The last Summary sentence says arrows for "fix graphics" - not sure what that means. Can arrows also be used for particles, e.g. as a dipole moment or spin vector ?

Also, please consider adding some of these nice example images to the dump image doc page, to illustrate the new capabilities.

akohlmey commented:
> @akohlmey The last Summary sentence says arrows for "fix graphics" - not sure what that means. Can arrows also be used for particles, e.g. as a dipole moment or spin vector ?

@sjplimp Please see the next pull request.

> Also, please consider adding some of these nice example images to the dump image doc page, to illustrate the new capabilities.

I was thinking about putting together a Howto page with commented examples. I want to get all my ideas implemented first and see where this takes me.

akohlmey commented:
> Also, please consider adding some of these nice example images to the dump image doc page, to illustrate the new capabilities.

@sjplimp
Hi Steve, I am working on this right now in PR #4820
I added three more image files to the "examples". But then I also 
moved a couple of "explanatory sections" from the dump image docs to Howto_viz.rst
and am adding some more discussions and examples to the howto instead of further
overloading the long doc page.

Please note that this pull request looks bigger than it is because it includes the changes from other pull requests that are not yet merged.

jrgissing commented on doc/src/dump_image.rst:
```diff
@@ -487,8 +495,52 @@ default the mapping of types to colors is as follows:
+   Since the rendered ellipsoids are constructed from iteratively refined
+   triangle meshes, the image quality increases with each refinement
+   level, but so does the computational effort to render the image.
+   Rendering only triangles is much faster than rendering the wireframe
```
Suggested rephrase:
Rendering only triangles is much faster than rendering the wireframe edges.  However, at mesh refinement levels of 4 and up, artifacts from the image rendering library can appear where triangles meet.  These artifacts can be somewhat hidden by using the *fsaa yes* setting, but are also less visible when rendering both edges and triangles.

akohlmey replied:
@jrgissing thanks. I've made the corresponding change in PR #4820 to minimize the chance of merge conflicts. I've also changed the statement a little bit, since I have since learned that those single pixel errors can happen whenever only triangles are used, but there are just so many fewer triangles a the lower levels and there are more pronounced edges, so they don't stand out as much.

jrgissing commented on doc/src/Howto_body.rst:
```diff
@@ -500,10 +500,11 @@ sphere is determined by the *bflag1* parameter for the *body* keyword.
-drawn diameter of each line segment is determined by the *bflag1*
-parameter for the *body* keyword.  The *bflag2* argument is ignored.
+and for *rounded/polyhedron* as a polyhedron.  The *bflag2* argument
+selects whether a wireframe mesh (2) is drawn or the faces (1) or both
```
are the values for this flag only described in this howto? I would recommend including values in the actual docs page. Also, can these values be changed to words instead of numbers?

akohlmey replied:
> I would recommend including values in the actual docs page.

The Howto is more-or-less the actual docs page. Most places in the other pages just have a reference to the Howto and state that it explains everything else.

> Also, can these values be changed to words instead of numbers?

Not without a major overhaul of the interface.

Please note that these are not my choices, but the way it already has been.
The interface for fixes to `dump image` is similarly awkward. For that I have introduced a new section for how `dump image` parameters are used to the fixes.

jrgissing commented on doc/src/fix_indent.rst:
```diff
@@ -223,15 +230,12 @@ and 3 means both ends are capped.
 object for spherical indenters, cylindrical indenters, and planar
-indenters in 2d systems.  In many cases you want to use a value < 0 to
-reduce the radius of the rendered object so that it does not obscure
+indenters in 2d systems.  In many cases you want to use a negative value
```
I suggest: A negative value may be necessary to reduce the radius of the rendered object so that it does not obscure nearby atoms.

jrgissing commented on doc/src/fix_smd_wall_surface.rst:
```diff
@@ -60,13 +60,19 @@ Fix *smd/wall\_surface* supports the *fix* keyword of :doc:`dump image
-transparency of the triangles and must use a value between 0.0
-(invisible) and 1.0 (fully opaque).  If using a mesh of cylinders, the
-*fflag2* setting determines the diameter of the cylinders.
+In case of using triangles, the *fflag2* setting is ignored.  If using a
```
I suggest: When rendering triangles, the *fflag2* setting is ignored.  When rendering a mesh of cylinders, the *fflag2* setting determines the diameter of the cylinders.

jrgissing reviewed (approved):
great set of additions!
